File: typescript/examples/workflow_wasm.js
import { Workflow, Client } from '../dist/index.js';

export async function run() {
  console.log("=== NativeBPM TS SDK: Workflow with WASM Service Task ===");

  // 1. Build workflow as code
  const workflow = new Workflow('wasm-demo', 'Workflow with WASM Service Task', './src/core.wasm');
  workflow.startEvent('start');
  
  // Add a Service Task configured with a WASM file path/alias.
  // The process engine will execute this compiled WASM binary inside a secure WASM sandbox.
  workflow.serviceTask('calculate', 'Calculate Totals', 'payment_topic')
    .wasm('./calculate_total.wasm');
  
  // Add an AI task that delegates decision checking to an LLM
  workflow.aiTask('aiCheck', 'AI Fraud Guard')
    .provider('google')
    .model('gemini-2.5-flash')
    .prompt('Analyze transaction for fraud: ${orderAmount}')
    .resultVar('isFraudulent');
      
  workflow.exclusiveGateway('gateway', 'Fraud Gateway');
  
  // Add a User (human) task for high-risk approval
  workflow.userTask('userTask', 'Manual Fraud Approval')
    .assignee('security_officer');
      
  workflow.endEvent('end', 'Process Finished');
  
  // Connect elements with sequence flows
  workflow.sequenceFlow('start', 'calculate');
  workflow.sequenceFlow('calculate', 'aiCheck');
  workflow.sequenceFlow('aiCheck', 'gateway');
  
  // Sequence flows with conditions
  workflow.sequenceFlowWithCondition('gateway', 'userTask', '${isFraudulent == true}');
  workflow.sequenceFlowWithCondition('gateway', 'end', '${isFraudulent == false}');
  workflow.sequenceFlow('userTask', 'end');
  
  // Compile the workflow AST to standard BPMN 2.0 XML using the embedded Go engine
  const bpmnXML = await workflow.buildXML();
  console.log("✓ Successfully compiled WASM workflow AST to BPMN 2.0 XML.");
  
  // 2. Deploy and start process definition using the Fluent Client API
  const client = new Client("http://localhost:8080", "test-bearer-token");
  
  console.log("\nDeploying to NativeBPM engine...");
  try {
    // Deploy process definition
    const definition = await client.definitions().deploy()
      .withID("wasm-demo")
      .withName("Workflow with WASM Service Task")
      .withBPMN(bpmnXML)
      .send();
    console.log(`✓ Deployed process definition (hash: ${definition.hash})`);
    
    // Start a process instance with input variables
    const instance = await client.instances().start("wasm-demo")
      .withBusinessKey("tx-8837")
      .withVariable("orderAmount", 2500)
      .send();
    console.log(`✓ Started process instance ID: ${instance.id} (completed: ${instance.completed})`);
    
  } catch (error) {
    console.log(`Note: Local API Engine deployment skipped (ensure local server is running on :8080). Details: ${error.message || error}`);
  }
}

// If run directly
if (import.meta.url.endsWith(process.argv[1])) {
  run();
}

File: python/examples/workflow_wasm.py
import os
from nativebpm import Client, Workflow

def main():
    print("=== NativeBPM Python SDK: Workflow with WASM Service Task ===")

    # 1. Build workflow as code using Fluent API method chaining
    workflow = Workflow('wasm-demo', 'Workflow with WASM Service Task', './nativebpm/core.wasm')
    
    # Chain starting from the start event
    workflow.start_event('start').next('calculate')
    
    # Configure WASM Service Task and chain to the AI task
    workflow.service_task('calculate', 'Calculate Totals', 'payment_topic')\
        .wasm('./calculate_total.wasm')\
        .next('aiCheck')
    
    # Configure AI task and chain to the exclusive gateway
    workflow.ai_task('aiCheck', 'AI Fraud Guard')\
        .provider('google')\
        .model('gemini-2.5-flash')\
        .prompt('Analyze transaction for fraud: ${orderAmount}')\
        .result_var('isFraudulent')\
        .next('gateway')
        
    # Set up exclusive gateway with transition conditions and default flow
    workflow.exclusive_gateway('gateway', 'Fraud Gateway')\
        .condition('userTask', '${isFraudulent == true}')\
        .default('end')
    
    # Configure User Task and chain to the end event
    workflow.user_task('userTask', 'Manual Fraud Approval')\
        .assignee('security_officer')\
        .next('end')
        
    workflow.end_event('end', 'Process Finished')
    
    # Compile the workflow AST to standard BPMN 2.0 XML using the embedded Go engine
    bpmn_xml = workflow.build_xml()
    print("✓ Successfully compiled WASM workflow AST to BPMN 2.0 XML.")
    
    # 2. Deploy and start process definition using the Fluent Client API
    client = Client("http://localhost:8080", "test-bearer-token")
    
    print("\nDeploying to NativeBPM engine...")
    try:
        # Deploy process definition
        definition = client.definitions().deploy()\
            .with_id("wasm-demo")\
            .with_name("Workflow with WASM Service Task")\
            .with_bpmn(bpmn_xml.encode('utf-8'))\
            .send()
        print(f"✓ Deployed process definition (hash: {definition.hash})")
        
        # Start a process instance with input variables
        instance = client.instances().start("wasm-demo")\
            .with_business_key("tx-8837")\
            .with_variable("orderAmount", 2500)\
            .send()
        print(f"✓ Started process instance ID: {instance.id} (completed: {instance.completed})")
        
    except Exception as e:
        print(f"Note: Local API Engine deployment skipped (ensure local server is running on :8080). Details: {e}")

if __name__ == '__main__':
    main()
